### PR TITLE
Backslash escapes

### PR DESCRIPTION
**Pull request by steve-h**
Reworked a lot of backslash handling for CommonMark spec conformance.

I am leaving a breaking test with escapes in a URL that requires discussion on what to do with this case of escaped text with a context like URL.  The reader cannot be advanced as in the existing code since the backslash escapes can change length of substring. 

~I also have not researched if "\<" in this context becomes "<" or "$lt;"~

I have added the CommonMark URL escape as implemented by cMark.

### CODE REVIEW

**Comment by steve-h** [comment]
I think it is pretty important to finish off the escaping stuff as it would potentially introduce many subtle changes to all users output as a "breaking" change in the future.
There are some more test to add but I needed to stop at the URL issue for some discussion of what code utility to add to fix the one breaking test case that is deliberately failing so far.

**Comment by JohnSundell** [comment]
Looks like the tests are failing @steve-h, let me know if you can't access the Bitrise logs for debugging.

**Comment by steve-h** [comment]
Yes I submitted a deliberately broken test to force you to think how you wish me to code the solution to it.

**Comment by JohnSundell** [comment]
Oh I see, I completely missed that 😅 Sorry it's really late here. Will take a look first thing in the morning 👍 Merged all of your other PRs for now, so we're making progress 😀

File: Sources/Ink/Internal/Character+Escaping.swift [hunk 1]
```diff
@@ -3,8 +3,80 @@
...
 */
+// ASCII chars escapable from CommonMark spec
+// \!\"\#\$\%\&\'\(\)\*\+\,\-\.\/\:\;\<\=\>\?\@\[\\\]\^\_\`\{\|\}\~
+// Result desired
+// !&quot;#$%&amp;'()*+,-./:;&lt;=&gt;?@[\]^_`{|}~
+// An ASCII punctuation character is !, ", #, $, %, &, ', (, ), *, +, ,, -, .,/
+// (U+0021–2F), :, ;, <, =, >, ?, @ (U+003A–0040), [, \, ], ^, _,`
+// (U+005B–0060), {, |, }, or ~ (U+007B–007E).
+let escapeSubstitutions: [Character: String] = [
+    "!": "!",
```
Comment (JohnSundell) [inline Sources/Ink/Internal/Character+Escaping.swift]: Why do we need entries for characters that are not being substituted, where the key and value are the exact same (which seems to be the majority of entries)? 🤔

File: Sources/Ink/Internal/Character+Escaping.swift [hunk 2]
```diff
@@ -3,8 +3,80 @@
...
+    "}": "}",
+    "|": "|",
+    "~": "~",
+    ">": "&gt;",
+    "<": "&lt;",
+    "&": "&amp;",
+    "\"": "&quot;",
+    Character(Unicode.Scalar(0000)): String(Unicode.Scalar(UInt32(0xFFFD))!),
+    //Took some research to initialize this one. Security of null char replaced so it cannot be used to cause end of file/string problems somewhere else. I don't think this is easy to test?
+    Character(Unicode.Scalar(UInt32(0x0021))): String(Unicode.Scalar(UInt32(0x0021))!),
```
Comment (JohnSundell) [inline Sources/Ink/Internal/Character+Escaping.swift]: Is it just me or does this seem a bit like overkill? Is the benefit of including this worth the effort of understanding/maintaining all of these manual unicode codes? Also seems like they are mostly `Character: String` pairs of the same exact character code?

File: Sources/Ink/Internal/Character+Escaping.swift [hunk 3]
```diff
@@ -3,8 +3,80 @@
...
+    Character(Unicode.Scalar(UInt32(0x007C))): String(Unicode.Scalar(UInt32(0x007C))!),
+    Character(Unicode.Scalar(UInt32(0x007D))): String(Unicode.Scalar(UInt32(0x007D))!),
+    Character(Unicode.Scalar(UInt32(0x007E))): String(Unicode.Scalar(UInt32(0x007E))!),
+    ]
+
+func escaped(_ char: Character) -> String? { escapeSubstitutions[char] }
+
 
 internal extension Character {
+    @available(*, deprecated, message: "use escaped(_ char: Character) -> String?")
```
Comment (JohnSundell) [inline Sources/Ink/Internal/Character+Escaping.swift]: We don't need to use deprecations on internal APIs. Just replace the call site.

File: Sources/Ink/Internal/Readable.swift [hunk 4]
```diff
@@ -10,10 +10,10 @@ internal protocol Readable {
...
 extension Readable {
     static func readOrRewind(using reader: inout Reader) throws -> Self {
-        guard reader.previousCharacter != "\\" else {
-            throw Reader.Error()
-        }
-
+//        guard reader.previousCharacter != "\\" else {
+//            throw Reader.Error()
+//        }
+// Not sure why this is here it is interfering with proper backslash handling
```
Comment (JohnSundell) [inline Sources/Ink/Internal/Readable.swift]: To enable top-level fragments, like lists and headings to be escaped.

**Comment by JohnSundell** [comment]
Also cc @ezfe and @john-mueller

File: Sources/Ink/Internal/Character+Escaping.swift (hunk 1 above)
Comment (steve-h) [reply Sources/Ink/Internal/Character+Escaping.swift]: This is not obvious and certainly deserves better comment to explain.
The amazing spec is that escaped CHARACTERS are not escaped (look at following test case, testOtherCharactersNotEscaped) The easiest way to make that work is to exhaustively handle ALL the specified punctuation chars at this dict lookup.

File: Sources/Ink/Internal/Readable.swift (hunk 4 above)
Comment (steve-h) [reply Sources/Ink/Internal/Readable.swift]: Well the best argument is that the escaping behaviour change is quite subtle and breaking to make later rather than earlier.  The history of all the handling of escapes is in the evolution of CommonMark. I am not a zealot here; but I respect John MacFarlane, author of pandoc, as he built the ultimate markdown rosetta tool and knows all the divergences.

I know is makes the code less pretty. I am thinking of it as a swift challenge test to match one of the premier libraries Cmark. These little items could make the onboarding experience of all other markdown files nasty.

The only other way to solve it is to use pandoc as a preprocessor to change CommonMark to Ink so you would have to brush up on haskell :)

I am just pounding out these backslash tests and have not done lists etc. yet.

File: Sources/Ink/Internal/Character+Escaping.swift (hunk 3 above)
Comment (steve-h) [reply Sources/Ink/Internal/Character+Escaping.swift]: I agree. I did it for myself so as to not miss places to check. There were some pending PRs that were in the area using this code and leaving it was safer. I am going to take it out now. I think the dictionary should be faster than the switch but I did not benchmark.

File: Sources/Ink/Internal/Character+Escaping.swift (hunk 1 above)
Comment (steve-h) [reply Sources/Ink/Internal/Character+Escaping.swift]: I cleaned it up a little as I misunderstood the spec. The Unicode hex stuff was not extra characters; just an explanation of the preceding ones. They should have used a semicolon since it was already comma separated.

**Comment by john-mueller** [comment]
I'll try to give this an extra pair of eyes over the weekend.

**Comment by steve-h** [comment]
I removed my failing test as it looks like a much more elaborate Link parser is needed for CommonMark.  I tried a trimmed down version of FormattedText but total failure and in reading the spec it looks like enough custom cases to require a custom parser. How it got so complicated is beyond my knowledge.  I have solved many backslash escapes that maybe can be merged as a subset.

File: Sources/Ink/Internal/Character+Escaping.swift (hunk 1 above)
Comment (steve-h) [reply Sources/Ink/Internal/Character+Escaping.swift]: After further playing with code blocks testing, I had to reintroduce the original HTML only escaping for use inside code and the punctuation testing escape for other cases.

**Comment by JohnSundell** [comment]
So sorry for the delay on reviewing this, and I'm afraid that it's going to be a few days more 🙁 I'm currently in the process of moving to a new city, so have had almost zero time for open source lately. But, I'm almost done, so next week I'll be able to review this (and hopefully merge it in). Again, apologies, and I really appreciate your work on this.

**Comment by steve-h** [comment]
I am really wanting you to release Publish as a priority.  Moving needs to happen too.  I am fine waiting.  I have been doing more research on OWASP cross site scripting prevention and other frameworks, like Go, and how they are handling escapes. I'm not even sure CommonMark has is right!  Any updates I keep submitting are probably because of these new learnings; so I am not sure the merge is ready anyways.

**Comment by steve-h** [comment]
**Don't merge yet!**  I need to move some code around.

I realized that the original place where the HTML escaping was done is wrong; it needs to be moved to the html output phase.  The reason is that modifiers are not applied until then and the modifier would probably want to see the raw text without having to unescape it.

Also the output of the Modifier should also be html-escaped to make sure it is a valid Markdown result.

Likewise href url escaping should be at the html output stage.

The backslash before punctuation logic is still valid but I need to decouple it from HTML escape.

This will also help if XML output is needed in the future where we can emit <br /> and the extra escaping of a single quote.

File: Sources/Ink/Internal/Character+Escaping.swift (hunk 2 above)
Comment (steve-h) [reply Sources/Ink/Internal/Character+Escaping.swift]: I reworked the  code so that an escaped punctuation causes the parser to skip over the backslash and then continue to use the "Substring of the original source strategy" for the retained punctuation. The tricky bit is to not allow the punctuation to then be parsed as a Markdown syntax character. Testcases should help this from regressing.

**Comment by steve-h** [comment]
This PR will be separated. The backslash of punctuation is a commonmark spec issue not necessarily needed. The html escaping should occur at html output and not during parse. The URL escaping is not needed for simple websites with no spaces in filenames. Publish fixes the spaces before publishing file paths.

The NULL character defence should be in Publish and could also be solved by grooming input if necessary since these characters will not be accidentally introduced in swift code. Commonmark whitespace is also just a conformance issue and whitespace is fixed by offline html minimizers that should be run a the tail end of the publishing when the files are gzipped for faster delivery.

**Comment by JohnSundell** [comment]
Sounds great! I'd love to see PRs making those kind of small improvements - those are super easy to test, review and merge in 👍 Also 100% agree on focusing on solving real-world issues first, then we can focus on improving edge-case conformances to CommonMark at a later stage.
